open_spiel/games/cheat.cc

```diff
@@ -0,0 +1,620 @@
+  // Did anyone shoot the moon?
+  Player potential_shooter = kInvalidPlayer;
+  bool moon_shot = true;
+  for (int i = 0; i < kNumTricks; ++i) {
+    int points = tricks_[i].Points();
+    // JD not required to shoot the moon.
+    if (points != 0 && points != kPointsForJD) {
+      // This trick must be taken by the shooter.
+      if (potential_shooter == kInvalidPlayer) {
+        potential_shooter = tricks_[i].Winner();
+      } else if (potential_shooter != tricks_[i].Winner()) {
+        moon_shot = false;
+        break;
+      }
+    }
+  }
+  // Shooting the moon sets the shooter's points to 0, and adds 26 pts to each
+  // opponent's score.
+  if (moon_shot) {
+    for (int i = 0; i < kNumPlayers; ++i) {
+      points_[i] += (i == potential_shooter) ? -kTotalPositivePoints
+                                             : kTotalPositivePoints;
+    }
+  }
+  // Did anyone avoid taking any tricks?
+  if (avoid_all_tricks_bonus_ && !moon_shot) {
+    std::vector<int> tricks_taken(kNumPlayers, 0);
+    for (int i = 0; i < kNumTricks; ++i) {
+      tricks_taken[tricks_[i].Winner()] += 1;
+    }
+    for (int i = 0; i < kNumPlayers; ++i) {
+      if (tricks_taken[i] == 0) points_[i] += kAvoidAllTricksBonus;
+    }
+  }
+}
+
+std::vector<double> CheatState::Returns() const {
+  if (!IsTerminal()) {
+    return std::vector<double>(kNumPlayers, 0.0);
+  }
+  std::vector<double> returns = points_;
+  for (int i = 0; i < returns.size(); ++i)
+    returns[i] = kTotalPositivePoints - returns[i];
+  return returns;
+}
+
+absl::optional<Player> CheatState::Played(int card) const {
+  if (phase_ == Phase::kPlay && !player_hand_[card].has_value()) {
+    Player p = *(initial_deal_[card]);
+    // check if they kept the card or not
+    auto it = std::find(passed_cards_[p].begin(), passed_cards_[p].end(), card);
+    if (it != passed_cards_[p].end()) {
+      p = (p + static_cast<int>(pass_dir_)) % kNumPlayers;
+    }
+    return p;
+  }
+
+  return absl::nullopt;
+}
+
+bool CheatState::KnowsLocation(Player player, int card) const {
+  bool dealt, received, played, two_clubs;
+  dealt = initial_deal_[card] == player;
+  int pass_dir = static_cast<int>(pass_dir_);
+  Player recv_from = (player + kNumPlayers - pass_dir) % kNumPlayers;
+  auto it = std::find(passed_cards_[recv_from].begin(),
+                      passed_cards_[recv_from].end(), card);
+  received = it != passed_cards_[recv_from].end() && phase_ == Phase::kPlay;
+  played = !player_hand_[card].has_value() && phase_ == Phase::kPlay;
+  two_clubs = card == Card(Suit::kClubs, 0) && phase_ == Phase::kPlay;
+  return dealt || received || played || two_clubs;
+}
+
+// Does not account for void suit information exposed by other players during
+// the play phase
+// std::unique_ptr<State> CheatState::ResampleFromInfostate(
+//     int player_id, std::function<double()> rng) const {
+//   std::unique_ptr<State> clone = game_->NewInitialState();
+//   Action pass_dir = static_cast<int>(pass_dir_);
```

Comment:
A lot of commented out code in this file. Did you intend to submit this code as a PR to add to OpenSpiel? If so, please remove all the commented code.